base/compiler/optimize.jl

```diff
@@ -453,9 +457,262 @@ function finish(interp::AbstractInterpreter, opt::OptimizationState,
+        if cfg !== nothing
+            return (cfg, domtree)
+        end
+        cfg = copy(ir.cfg)
+        # Add a virtual basic block to represent the exit
+        push!(cfg.blocks, BasicBlock(StmtRange(0:-1)))
+
+        for bb = 1:(length(cfg.blocks)-1)
+            terminator = ir[SSAValue(last(cfg.blocks[bb].stmts))][:inst]
+            if isa(terminator, ReturnNode) && isdefined(terminator, :val)
+                cfg_insert_edge!(cfg, bb, length(cfg.blocks))
+            end
+        end
+
+        domtree = construct_domtree(cfg.blocks)
+        return (cfg, domtree)
+    end
+
+    function is_getfield_boundscheck(inst::Instruction)
+        stmt = inst[:stmt]
+        is_known_call(stmt, getfield, ir) || return false
+        length(stmt.args) < 4 && return false
+        boundscheck = stmt.args[end]
+        argextype(boundscheck, ir) === Bool || return false
+        isa(boundscheck, SSAValue) || return false
+        return true
+    end
+
+    function is_conditional_noub(inst::Instruction)
+        # Special case: `:boundscheck` into `getfield`
+        is_getfield_boundscheck(inst) || return false
+        barg = inst[:stmt].args[end]
+        isexpr(ir[barg][:stmt], :boundscheck) || return false
+        # If IR_FLAG_INBOUNDS is already set, no more conditional ub
+        (ir[barg][:flag] & IR_FLAG_INBOUNDS) != 0 && return false
+        any_conditional_ub = true
+        return true
+    end
+
+    function scan_non_dataflow_flags!(inst::Instruction)
+        flag = inst[:flag]
+        all_effect_free &= (flag & IR_FLAG_EFFECT_FREE) != 0
+        all_nothrow &= (flag & IR_FLAG_NOTHROW) != 0
+        if (flag & IR_FLAG_NOUB) == 0
+            if !is_conditional_noub(inst)
+                all_noub = false
+            end
+        end
+    end
+
+    function scan_inconsistency!(inst::Instruction, idx)
+        flag = inst[:flag]
+        stmt_inconsistent = (flag & IR_FLAG_CONSISTENT) == 0
+        stmt = inst[:stmt]
+        # Special case: For getfield, we allow inconsistency of the :boundscheck argument
+        if is_getfield_boundscheck(inst)
+            for i = 1:(length(stmt.args)-1)
+                val = stmt.args[i]
+                if isa(val, SSAValue)
+                    stmt_inconsistent |= val.id in inconsistent
+                    count!(tpdum, val)
+                end
+            end
+        else
+            for ur in userefs(stmt)
+                val = ur[]
+                if isa(val, SSAValue)
+                    stmt_inconsistent |= val.id in inconsistent
+                    count!(tpdum, val)
+                end
+            end
+        end
+        stmt_inconsistent && push!(inconsistent, idx)
+        return stmt_inconsistent
+    end
+
+    function scan_stmt!(inst, idx, lstmt, bb)
+        stmt = inst[:inst]
+        flag = inst[:flag]
+
+        if isexpr(stmt, :enter)
+            # try/catch not yet modeled
+            had_trycatch = true
+            return true
```

Comment:
```suggestion
            return false
```
Shouldn't return `false` to bail out the scan?